docs/source/tutorial-pyfunc.rst

```diff
@@ -0,0 +1,368 @@
+    from sklearn.metrics import accuracy_score
+
+    import pandas as pd
+    import joblib as jl
+    import numpy as np
+
+    import mlflow
+
+    from custom_code import iris_classes
+
+
+    def build_and_train_model(X : pd.DataFrame, y : pd.Series) -> LogisticRegression:
+        classifier = LogisticRegression(C=1.0, random_state=42).fit(X, y)
+        return classifier
+
+
+    def predict_and_track(classifier : LogisticRegression, X : pd.DataFrame, y : pd.Series) -> float:
+        predictions = classifier.predict(X)
+        score = accuracy_score(y, predictions)
+        return score
+
+
+    def initialize_and_train() -> (LogisticRegression, ModelSignature):
+        X, y = load_iris(return_X_y=True, as_frame=True)
+        classifier = build_and_train_model(X, y)
+        score = predict_and_track(classifier, X, y)
+        mlflow.log_metric("accuracy", score)
+        signature = mlflow.models.infer_signature(X, y)
+        return classifier, signature
+
+
+        # Entry point
+        if __name__ == "__main__":
+            with mlflow.start_run() as run:
+                classifier, signature = initialize_and_train()
+
+Once you've created the file, run it with:
+
+.. code-block:: bash
+
+    python train.py
+
+.. note:: You may receive a warning that the solver failed to converge. However, if you check the accuracy of ``predictions`` vs ``y``, you will see that the model is highly accurate.
+
+The code:
+
+1. Imports necessary modules from ``sklearn`` and ``mlflow``. 
+2. Starts an MLflow run (``with mlflow.start_run() as run``)
+3. Within ``initialize_and_train()``: loads the iris dataset 
+4. Specifies and fits a ``LogisticRegression`` model (``build_and_train_model``)
+5. Calculates the accuracy of the model and logs it to MLflow (``predict_and_track``)
+6. Infers the signature of the model using :py:func:`mlflow.models.infer_signature` function and the input data and labels
+7. Returns the model and the signature of the model
+
+What it doesn't do is log the model! Because there is an sklearn flavor for MLflow, we could do so by adding a single line of code to main:
+
+.. code-block:: python
+
+    mlflow.sklearn.log_model(classifier, "model", signature=signature)
+
+However, for the purposes of this tutorial, we'll create a custom model and log it using ``mlflow.pyfunc.log_model``. 
+
+Create a Custom ``pyfunc`` Model
+--------------------------------
+
+To create a custom ``pyfunc`` model, you need to create a class that inherits from ``mlflow.pyfunc.PythonModel``. This class must implement two methods: ``load_context`` and ``predict``. The ``load_context`` method is called when the model is loaded. It takes a single argument, ``context`` and prepares the model for inferencing. The ``predict`` method is called when the model is used for inference. It takes two arguments: ``context`` and ``model_input``. The ``model_input`` argument is the input data for the model. The ``predict`` method returns the model's predictions.
```

Comment:
alternatively, declaring a single function named `predict` that contains the logic needed to perform the prediction on an arbitrary model is supported. if such a function is passed in to the `python_model` argument of `mlflow.pyfunc.log_model()`, we assemble the PyFunc class on behalf of the user, inserting their defined function as a method in the Class skeleton. We should mention this behavior.

Reply:
of note: in order to use this functionality, the `artifacts` dict must be specified. That should be explicitly called out below, when talking about the function of that argument (single purpose, dual use w.r.t. the dependency that partial function passing has on the fact that it is provided)